File: 91.解码方法.py
#
# @lc app=leetcode.cn id=91 lang=python3
#
# [91] 解码方法
#

# @lc code=start
NUM = 0

def search(s: str) -> int:
    global NUM
    if s == "" :
        NUM += 1
        return 1
    num1 = int(s[0:1])
    if num1 > 0:
        search(s[1:])
    else:
        return 0
    if len(s) < 2:
        return
    num2 = int(s[0:2])
    if num2 > 0 and num2 < 27:
        search(s[2:])
    return 0

D = []
def decode(l: int) -> int:
    global D
    if l ==1 or l == 0:
        return 1
    if D[l] != -1:
        return D[l]
    D[l] = decode(l-1) + decode(l-2)
    return D[l]

def split(s: str):
    pos = [0]
    rs = []
    l = 0
    i = 0
    while i < len(s):
        l += 1
        if s[i] == '0':
            if l-2 < 0:
                return []
            if i >=2:
                rs.append(l-2)
                rs.append(1)
                l = 0
                i += 1
                continue
            elif i == 0:
                return []
            elif i == 1:
                rs.append(1)
                l = 0
                i += 1
                continue
        if i<len(s)-1 and int(s[i:i+2]) > 26:
            rs.append(l)
            l = 0
        i += 1
    if l != 0:
        rs.append(l)
    return rs

class Solution:
    def numDecodings(self, s: str) -> int:
        lens = split(s)
        print(lens)
        if len(lens) == 0:
            return 0
        m = max(lens)
        global D
        for _ in range(m+2):
            D.append(-1)
        rs = 1
        for l in lens:
            rs = rs * decode(l)
        return rs
# @lc code=end
if __name__ == "__main__":
    s = Solution()
    print(s.numDecodings("201"))

File: 91.解码方法.cpp
/*
 * @lc app=leetcode.cn id=91 lang=cpp
 *
 * [91] 解码方法
 *
 * https://leetcode-cn.com/problems/decode-ways/description/
 *
 * algorithms
 * Medium (30.64%)
 * Likes:    955
 * Dislikes: 0
 * Total Accepted:    162K
 * Total Submissions: 526.4K
 * Testcase Example:  '"12"'
 *
 * 一条包含字母 A-Z 的消息通过以下映射进行了 编码 ：
 * 
 * 
 * 'A' -> 1
 * 'B' -> 2
 * ...
 * 'Z' -> 26
 * 
 * 
 * 要 解码 已编码的消息，所有数字必须基于上述映射的方法，反向映射回字母（可能有多种方法）。例如，"11106" 可以映射为：
 * 
 * 
 * "AAJF" ，将消息分组为 (1 1 10 6)
 * "KJF" ，将消息分组为 (11 10 6)
 * 
 * 
 * 注意，消息不能分组为  (1 11 06) ，因为 "06" 不能映射为 "F" ，这是由于 "6" 和 "06" 在映射中并不等价。
 * 
 * 给你一个只含数字的 非空 字符串 s ，请计算并返回 解码 方法的 总数 。
 * 
 * 题目数据保证答案肯定是一个 32 位 的整数。
 * 
 * 
 * 
 * 示例 1：
 * 
 * 
 * 输入：s = "12"
 * 输出：2
 * 解释：它可以解码为 "AB"（1 2）或者 "L"（12）。
 * 
 * 
 * 示例 2：
 * 
 * 
 * 输入：s = "226"
 * 输出：3
 * 解释：它可以解码为 "BZ" (2 26), "VF" (22 6), 或者 "BBF" (2 2 6) 。
 * 
 * 
 * 示例 3：
 * 
 * 
 * 输入：s = "0"
 * 输出：0
 * 解释：没有字符映射到以 0 开头的数字。
 * 含有 0 的有效映射是 'J' -> "10" 和 'T'-> "20" 。
 * 由于没有字符，因此没有有效的方法对此进行解码，因为所有数字都需要映射。
 * 
 * 
 * 示例 4：
 * 
 * 
 * 输入：s = "06"
 * 输出：0
 * 解释："06" 不能映射到 "F" ，因为字符串含有前导 0（"6" 和 "06" 在映射中并不等价）。
 * 
 * 
 * 
 * 提示：
 * 
 * 
 * 1 
 * s 只包含数字，并且可能包含前导零。
 * 
 * 
 */

#include <string>
using namespace std;
// @lc code=start
class Solution {
public:
    int numDecodings(string s) {
        int dp[100] = {};
        dp[0] = 0;
        if (s[0] == '0') {
            return 0;
        }
        dp[1] = 1;
        for (size_t i = 1; i < s.length(); i++) {
            int di = i + 1;
            if (s[i] == '0') {
                dp[di] = 0;
            } else {
                dp[di] = dp[di-1];
            }
            if (s[i-1] <= '2' && s[i] <= '6') {
                dp[di] += dp[di-2];
            } 
            if (dp[di] == 0) {
                return 0;
            }
        }
        return dp[s.length()];
    }
};
// @lc code=end

